include/ginkgo/core/distributed/index_map.hpp

```diff
@@ -0,0 +1,180 @@
+ * separated into locally owned (index_space::local) and non-locally owned
+ * (index_space::non_local). The locally owned indices are defined as
+ * $[0, \dots, |I_k|)$, and the non-locally owned as $[0, \dots, |R_k|)$.
+ * With these index sets, the following mappings are defined:
+ * - $c_k : \hat{I}_k \mapsto [0, \dots, |\hat{I}_k|)$ which maps global indices
+ *   into the combined/full local index space,
+ * - $l_k: I_k \mapsto [0, \dots, |I_k|)$ which maps global indices into the
+ *   locally owned index space,
+ * - $r_k: R_k \mapsto [0, \dots, |R_k|)$ which maps global indices into the
+ *   non-locally owned index space.
+ * The required map can be selected by passing the appropriate type of an
+ * index_space.
+ *
+ * The index map for $I_k$ has no knowledge about any other index maps for
+ * $I_l, l \neq k$. In particular, any global index passed to the `get_local`
+ * map that is not part of the specified index space, will be mapped to an
+ * invalid_index.
+ *
+ * \tparam LocalIndexType  type for local indices
+ * \tparam GlobalIndexType  type for global indices
+ */
+template <typename LocalIndexType, typename GlobalIndexType = int64>
+struct index_map {
+    using part_type = Partition<LocalIndexType, GlobalIndexType>;
+
+    /**
+     * \brief Maps global indices to local indices
+     *
+     * \param global_ids  the global indices to map
+     * \param index_space_v  the index space in which the returned local indices
+     *                       are defined
+     *
+     * \return  the mapped local indices. Any global index that is not in the
+     *          specified index space is mapped to invalid_index.
+     */
+    [[nodiscard]] array<LocalIndexType> get_local(
+        const array<GlobalIndexType>& global_ids,
+        index_space index_space_v = index_space::combined) const;
+
+    /**
+     * \brief get size of index_space::local
+     */
+    [[nodiscard]] size_type get_local_size() const
+    {
+        return partition_ ? partition_->get_part_size(rank_) : 0;
+    }
+
+    /**
+     * \brief get size of index_space::non_local
+     */
+    [[nodiscard]] size_type get_non_local_size() const
+    {
+        return remote_global_idxs_.get_flat().get_size();
+    }
+
+    /**
+     * \brief Creates a new index map.
+     *
+     * The passed in recv_connections may contain duplicates, which will be
+     * filtered out.
+     *
+     * \param exec  the executor
+     * \param part  the partition of the global index set
+     * \param rank  the id of the global index space subset
+     * \param recv_connections  the global indices that are not owned by this
+     *                          rank, but accessed by it
+     */
+    index_map(std::shared_ptr<const Executor> exec,
+              std::shared_ptr<const part_type> part, comm_index_type rank,
+              const array<GlobalIndexType>& recv_connections);
+
+    /**
+     * \brief Creates an empty index map.
+     */
+    index_map() = default;
```

Comment:
What case do you have in mind for an empty index_map? Its executor and its arrays' executors are uninitialized, that seems problematic

Reply:
I think a empty index map is a valid object, so I would like to keep it. I guess I could require an executor, that wouldn't change much.

Reply:
This is gone now, it's replaced by a constructor with a single executor.